From 35cd9d1e2fc0de25a5dadcb7599ebc6cf485f4df Mon Sep 17 00:00:00 2001
From: Aleksei Greshilov <aleksei.greshilov@cern.ch>
Date: Mon, 2 Sep 2024 19:00:45 -0400
Subject: [PATCH] DMA Proxy driver fix for X2O

---
 drivers/Kconfig  | 4 ++--
 drivers/Makefile | 6 +++---
 2 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/drivers/Kconfig b/drivers/Kconfig
index 1fd9dc570385..4d7ac4c0f277 100644
--- a/drivers/Kconfig
+++ b/drivers/Kconfig
@@ -3,8 +3,6 @@ menu "Device Drivers"
 
 # Keep I/O buses first
 
-source "drivers/x2o/Kconfig"
-
 source "drivers/amba/Kconfig"
 source "drivers/eisa/Kconfig"
 source "drivers/pci/Kconfig"
@@ -12,6 +10,8 @@ source "drivers/cxl/Kconfig"
 source "drivers/pcmcia/Kconfig"
 source "drivers/rapidio/Kconfig"
 
+source "drivers/x2o/Kconfig"
+
 source "drivers/base/Kconfig"
 
 source "drivers/bus/Kconfig"
diff --git a/drivers/Makefile b/drivers/Makefile
index 7e42669b88da..e720161025d0 100644
--- a/drivers/Makefile
+++ b/drivers/Makefile
@@ -39,6 +39,9 @@ obj-y				+= clk/
 # really early.
 obj-$(CONFIG_DMADEVICES)	+= dma/
 
+# X2O platform drivers.
+obj-$(CONFIG_DMADEVICES)        += x2o/
+
 # SOC specific infrastructure drivers.
 obj-y				+= soc/
 
@@ -188,6 +191,3 @@ obj-$(CONFIG_GNSS)		+= gnss/
 obj-$(CONFIG_INTERCONNECT)	+= interconnect/
 obj-$(CONFIG_COUNTER)		+= counter/
 obj-$(CONFIG_MOST)		+= most/
-
-# X2O platform drivers.
-obj-$(CONFIG_DMADEVICES)        += x2o/
